src/promptflow/promptflow/_sdk/_record_storage.py

```diff
@@ -0,0 +1,228 @@
+        :type input_dict: OrderedDict
+        :raises RecordFileMissingException: Record file not exist
+        :raises RecordItemMissingException: Record item not exist in record file
+        :return: original output of node run
+        :rtype: object
+        """
+        hash_value: str = hashlib.sha1(str(input_dict).encode("utf-8")).hexdigest()
+        current_saved_records: Dict[str, str] = self.cached_items.get(self._record_file_str, None)
+        if current_saved_records is None:
+            raise RecordFileMissingException(
+                f"This should except earlier, record file not found in folder {self.input_path}."
+            )
+        saved_output = current_saved_records.get(hash_value, None)
+        if saved_output is None:
+            raise RecordItemMissingException(
+                f"Record item not found in folder {self.input_path}.\n" f"values: {json.dumps(input_dict)}\n"
+            )
+
+        # not all items are reserved in the output dict.
+        output = saved_output["output"]
+        output_type = saved_output["output_type"]
+        if isinstance(output, str) and output_type != "str":
+            output_convert = json.loads(output)
+            return output_convert
+        else:
+            return output
+
+    def set_record(self, input_dict: OrderedDict, output: object) -> None:
+        """
+        set_record Set record to local storage, always override the old record.
+
+        :param input_dict: input dict of critical AOAI inputs
+        :type input_dict: OrderedDict
+        :param output: original output of node run
+        :type output: object
+        """
+        hash_value: str = hashlib.sha1(str(input_dict).encode("utf-8")).hexdigest()
+        current_saved_records: Dict[str, str] = self.cached_items.get(self._record_file_str, None)
+
+        if current_saved_records is None:
+            current_saved_records = {}
+            current_saved_records[hash_value] = {
+                "input": input_dict,
+                "output": output,
+                "output_type": type(output).__name__,
+            }
+        else:
+            saved_output = current_saved_records.get(hash_value, None)
+            if saved_output is not None:
+                if saved_output["output"] == str(output) and saved_output["output_type"] == type(output).__name__:
+                    return
+                else:
+                    current_saved_records[hash_value] = {
+                        "input": input_dict,
+                        "output": output,
+                        "output_type": type(output).__name__,
+                    }
+            else:
+                current_saved_records[hash_value] = {
+                    "input": input_dict,
+                    "output": str(output),
+                    "output_type": type(output).__name__,
+                }
+        self.cached_items[self._record_file_str] = current_saved_records
+        self._write_file()
```

Comment:
I am wondering we might do append log line format here, to avoid writing full file if the recording gets bigger.

Reply:
Yes, of course. I could calculate the hash on the go.